container_api/src/interface.rs

```diff
@@ -156,10 +189,155 @@ impl ContainerApiBuilder {
+
+                let id = Self::get_as_string("id", &params_map)?;
+                let path = Self::get_as_string("path", &params_map)?;
+
+                container_call!(|c| c.install_dna_from_file(PathBuf::from(path), id.to_string()))?;
+
+                Ok(serde_json::Value::String("success".into()))
+            });
+
+        self.io.add_method("admin/dna/uninstall", move |params| {
+            let params_map = Self::unwrap_params_map(params)?;
+            let id = Self::get_as_string("id", &params_map)?;
+            container_call!(|c| c.uninstall_dna(&id))?;
+            Ok(serde_json::Value::String("success".into()))
+        });
+
+        self.io.add_method("admin/dna/list", move |_params| {
+            let dnas = container_call!(
+                |c| Ok(c.config.dnas.clone()) as Result<Vec<DnaConfiguration>, String>
+            )?;
+            Ok(serde_json::Value::Array(
+                dnas.iter()
+                    .map(|dna| json!({"id": dna.id, "hash": dna.hash}))
+                    .collect(),
+            ))
+        });
+
+        self.io.add_method("admin/instance/add", move |params| {
+            let params_map = Self::unwrap_params_map(params)?;
+
+            let id = Self::get_as_string("id", &params_map)?;
+            let dna_id = Self::get_as_string("dna_id", &params_map)?;
+            let agent_id = Self::get_as_string("agent_id", &params_map)?;
+
+            let new_instance = InstanceConfiguration {
+                id: id.to_string(),
+                dna: dna_id.to_string(),
+                agent: agent_id.to_string(),
+                storage: StorageConfiguration::Memory, // TODO: don't actually use this. Have some idea of default store
+            };
+
+            container_call!(|c| c.add_instance(new_instance))?;
+
+            Ok(serde_json::Value::String("success".into()))
+        });
+
+        self.io.add_method("admin/instance/remove", move |params| {
+            let params_map = Self::unwrap_params_map(params)?;
+            let id = Self::get_as_string("id", &params_map)?;
+            container_call!(|c| c.remove_instance(&id))?;
+            Ok(serde_json::Value::String("success".into()))
+        });
+
+        self.io.add_method("admin/instance/start", move |params| {
+            let params_map = Self::unwrap_params_map(params)?;
+            let id = Self::get_as_string("id", &params_map)?;
+            container_call!(|c| c.start_instance(&id))?;
+            Ok(serde_json::Value::String("success".into()))
+        });
+
+        self.io.add_method("admin/instance/stop", move |params| {
+            let params_map = Self::unwrap_params_map(params)?;
+            let id = Self::get_as_string("id", &params_map)?;
+            container_call!(|c| c.stop_instance(&id))?;
+            Ok(serde_json::Value::String("success".into()))
+        });
+
+        self.io.add_method("admin/instance/list", move |_params| {
```

Comment:
Feel free to remove `info/instances` if it hasn't been already, I think this is pretty much the same.

Although, `info/instances` needs to be accessible from every UI, i.e. without admin privileges, so this should be taken out of the admin namespace?

Reply:
I first mistook those two for the same too, but actually `info/instances` is something different:

* Yes, `info/instances` needs to be accessible in every interfaces as it lists all the instances available here.
* `admin/instance/list` is only available to admins and lists all instances that are configured.

Reply:
Ahh yes OK